pkg/executor/aggregate/agg_hash_final_worker.go

```diff
@@ -174,19 +213,192 @@ func (w *HashAggFinalWorker) receiveFinalResultHolder() (*chunk.Chunk, bool) {
+			if err != nil {
+				return totalMemDelta, err
+			}
+			totalMemDelta += memDelta
 		}
-		if w.stats != nil {
-			w.stats.WorkerTime += int64(time.Since(start))
+	} else {
+		totalMemDelta += int64(len(key))
+
+		if len(*context.restoreadData)+1 > (1<<*context.bInMap)*hack.LoadFactorNum/hack.LoadFactorDen {
+			w.memTracker.Consume(hack.DefBucketMemoryUsageForMapStrToSlice * (1 << *context.bInMap))
+			(*context.bInMap)++
+		}
+
+		results := make([]aggfuncs.PartialResult, context.aggFuncNum)
+		(*context.restoreadData)[key] = results
+
+		for aggPos := 0; aggPos < context.aggFuncNum; aggPos++ {
+			results[aggPos] = context.partialResultsRestored[aggPos][context.rowPos]
+		}
+	}
+	return totalMemDelta, nil
+}
+
+func (w *HashAggFinalWorker) restoreFromOneSpillFile(ctx sessionctx.Context, restoreadData *aggfuncs.AggPartialResultMapper, diskIO *chunk.DataInDiskByChunks, bInMap *int) (int64, error) {
+	totalMemDelta := int64(0)
+	chunkNum := diskIO.NumChunks()
+	aggFuncNum := len(w.aggFuncsForRestoring)
+	processRowContext := &processRowContext{
+		ctx:                    ctx,
+		chunk:                  nil, // Will be set in the loop
+		rowPos:                 0,   // Will be set in the loop
+		keyColPos:              aggFuncNum,
+		aggFuncNum:             aggFuncNum,
+		restoreadData:          restoreadData,
+		partialResultsRestored: make([][]aggfuncs.PartialResult, aggFuncNum),
+		bInMap:                 bInMap,
+	}
+	for i := 0; i < chunkNum; i++ {
+		chunk, err := diskIO.GetChunk(i)
+		if err != nil {
+			return totalMemDelta, err
+		}
+
+		// Deserialize bytes to agg function's meta data
+		for aggPos, aggFunc := range w.aggFuncsForRestoring {
+			partialResult, memDelta := aggFunc.DeserializePartialResult(chunk)
+			processRowContext.partialResultsRestored[aggPos] = partialResult
+			totalMemDelta += memDelta
+		}
+
+		// Merge or create results
+		rowNum := chunk.NumRows()
+		processRowContext.chunk = chunk
+		for rowPos := 0; rowPos < rowNum; rowPos++ {
+			processRowContext.rowPos = rowPos
+			memDelta, err := w.processRow(processRowContext)
+			if err != nil {
+				return totalMemDelta, err
+			}
+			totalMemDelta += memDelta
+		}
+	}
+	return totalMemDelta, nil
+}
+
+func (w *HashAggFinalWorker) restoreOnePartition(ctx sessionctx.Context) (bool, error) {
+	restoredData := make(aggfuncs.AggPartialResultMapper)
+	bInMap := 0
+
+	restoredPartitionIdx, isSuccess := w.spillHelper.getNextPartition()
+	if !isSuccess {
+		return false, nil
+	}
+
+	spilledFilesIO := w.spillHelper.getListInDisks(restoredPartitionIdx)
+	for _, spilledFile := range spilledFilesIO {
+		memDelta, err := w.restoreFromOneSpillFile(ctx, &restoredData, spilledFile, &bInMap)
+		if err != nil {
+			return false, err
 		}
-		waitGroup.Done()
-	}()
-	if err := w.consumeIntermData(ctx); err != nil {
+
+		w.memTracker.Consume(memDelta)
+	}
+	w.spilledDataChan <- &restoredData
```

Comment:
We do not need this channel, `restoredData` can be defined as a member attribute of spillHelper

Reply:
> We do not need this channel, `restoredData` can be defined as a member attribute of spillHelper

done